src/boot/z_std_dma.c

```diff
@@ -0,0 +1,470 @@
+        (size == 0) || (size % 2 != 0)) {
+        DmaMgr_Error(req, NULL, "ILLIGAL DMA-FUNCTION CALL", "パラメータ異常です");
+    }
+
+    osSendMesg(&sDmaMgrMsgQueue, req, OS_MESG_BLOCK);
+    return 0;
+}
+
+/**
+ * Submit a synchronous DMA request. This will block the current thread until the requested transfer is complete. Data
+ * is immediately available as soon as this function returns.
+ *
+ * @param vram Location in RAM for data to be written.
+ * @param vrom Virtual ROM location for data to be read.
+ * @param size Transfer size.
+ */
+s32 DmaMgr_RequestSync(void* vram, RomOffset vrom, size_t size) {
+    DmaRequest req;
+    OSMesgQueue mq;
+    OSMesg msg[1];
+    s32 ret;
+
+    osCreateMesgQueue(&mq, msg, ARRAY_COUNT(msg));
+    ret = DmaMgr_SendRequest(&req, vram, vrom, size, 0, &mq, NULL);
+    if (ret == -1) {
+        return ret;
+    }
+
+    osRecvMesg(&mq, NULL, OS_MESG_BLOCK);
+
+    return 0;
+}
+
+/**
+ * Synchronous DMA Request that calculates the size of the segment automatically.
+ *
+ * @see DmaMgr_RequestSync
+ */
+s32 DmaMgr_RequestSyncNoSize(void* vram, RomOffset vrom) {
+    DmaEntry* entry = DmaMgr_FindDmaEntry(vrom);
+
+    return DmaMgr_RequestSync(vram, vrom, entry->vromEnd - entry->vromStart);
+}
+
+/**
+ * Returns the uncompressed size for a given segment.
+ */
+size_t DmaMgr_GetSegmentSize(RomOffset vrom) {
+    DmaEntry* entry = DmaMgr_FindDmaEntry(vrom);
+
+    return entry->vromEnd - entry->vromStart;
+}
+
+/**
+ * Returns the ovl reloc vrom start for a given segment.
+ */
+RomOffset DmaMgr_GetOvlStart(RomOffset vrom) {
+    DmaEntry* entry = DmaMgr_FindDmaEntry(vrom);
+
+    return entry[1].vromStart;
+}
+
+/**
+ * Searches for the rom offsets of the ovl reloc segment for a given segment.
+ *
+ * The values will be stored in the pointers passed as parameters. If no segment matches the given vrom then the
+ * parameters will be left untouched.
+ *
+ * @param vromStart The vrom start of the segment.
+ * @param[out] vromEnd The vrom end for the given segment.
+ * @param[out] ovlStart The vrom start for the ovl reloc segment.
+ * @param[out] ovlEnd The vrom end for the ovl reloc segment.
+ * @return s32 0 if the given segment is found, -1 otherwise.
+ */
+s32 DmaMgr_GetOvlOffsets(RomOffset vromStart, RomOffset* vromEnd, RomOffset* ovlStart, RomOffset* ovlEnd) {
+    DmaEntry* entry = DmaMgr_FindDmaEntry(vromStart);
+
+    if (entry != NULL) {
+        *vromEnd = entry[0].vromEnd;
+        *ovlStart = entry[1].vromStart;
+        *ovlEnd = entry[1].vromEnd;
+        return 0;
+    }
+    return -1;
+}
+
+void DmaMgr_Init(void) {
+    // DMA the dma data table to RAM
+    DmaMgr_DmaRomToRam(SEGMENT_ROM_START(dmadata), SEGMENT_VRAM_START(dmadata), SEGMENT_ROM_SIZE(dmadata));
+
+    do {
+        DmaEntry* entry = dma_rom_ad;
+        s32 count = 0;
+
+        for (; entry->vromEnd != 0; entry++) {
+            count++;
+        }
+
+        sNumDmaEntries = count;
+    } while (0);
```

Comment:
Do you need the full `do while(0);`? Perhaps just scoped works (though not sure it is better). Should we have a `fake` label here?

Reply:
Doesn't need the `do while(0)` but it needs the scope. We usually don't do fake labels for scopes like this. Also, MM has this same scope https://github.com/zeldaret/mm/blob/cbf9d98dba61924625629061f7680e3508fc2701/src/boot_O2_g3/z_std_dma.c#L220-L230